### PR TITLE
Fix weight decay masking in `run_flax_glue.py`

### LINKED ISSUE #11936: Flax text-classification multi-optimizer incorrect
The below code is incorrect:

https://github.com/huggingface/transformers/blob/8d171628fe84bdf92ee40b5375d7265278180f14/examples/flax/text-classification/run_flax_glue.py#L173-L188

- The `dict` from `traverse_util.flatten_dict` has keys which are tuples of strings, not one long string with the path separated by periods.
- `optax.masked` applies the transformation wherever the mask is True, so the masks are flipped.
- Flax's LayerNorm calls the scale parameter `scale` not `weight`

I believe following code would be correct:

```py
# We use Optax's "masking" functionality to create a multi-optimizer, one
# with weight decay and the other without. Each sub-optimizer will be applied
# wherever the mask is True
decay_path = lambda p: p[-1] != "bias" and p[-2:] != ("LayerNorm", "scale")  # noqa: E731

tx = optax.chain(
    optax.masked(adamw(0.0), mask=traverse(lambda path, _: not decay_path(path))),
    optax.masked(adamw(weight_decay), mask=traverse(lambda path, _: decay_path(path))),
)
```

But, for the networks in that example, the parameters that shouldn't be decayed all have 1 dimension (biases and layernorm scales). L173-L188 can therefore be simplified to:

```py
# We use Optax's "masking" functionality to create a multi-optimizer, one
# with weight decay and the other without. Each sub-optimizer will be applied
# wherever the mask is True. The bias parameters and LayerNorm scale
# parameters should not be decayed, and these are the only parameters
# with 1 dimension.
tx = optax.chain(
    optax.masked(adamw(0.0), mask=partial(jax.tree_map, lambda p: p.ndim == 1)),
    optax.masked(adamw(weight_decay), mask=partial(jax.tree_map, lambda p: p.ndim != 1)),
)
```

Though the latter solution is simpler, perhaps the first one is better since it illustrates a more general way for users to construct a multi-optimizer. I'd be happy to open a PR with either.

cc: @marcvanzee @patrickvonplaten

### PR DESCRIPTION
**Pull request by n2cholas**
# What does this PR do?

Fixes #11936

In addition to the changes discussed in the issue, I combined `traverse` and `decay_path` into one function `decay_mask_fn` to simplify the implementation. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@patrickvonplaten

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
That's great - thanks a lot for the fix @n2cholas :-) I'll re-run the FlaxGlue suit today with your fix to check if results improve

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
Thanks a lot for the fix @n2cholas - I reran the eval train+eval script & updated the results accordingly
